cooked-validators/src/Cooked/Attack/Common.hs

```diff
@@ -1,234 +1,218 @@
+  pure x = Attack $ \_ skel -> [(skel, x)]
+  (<*>) = ap
+
+instance Monad Attack where
+  Attack g >>= h = Attack $ \mcst skel ->
+    concatMap (\(skel', x) -> getAttack (h x) mcst skel') $ g mcst skel
+
+instance Alternative Attack where
+  empty = Attack $ \_ _ -> []
+  Attack f <|> Attack g = Attack $ \mcst skel -> f mcst skel ++ g mcst skel
+
+instance MonadPlus Attack
+
+-- * Simple attacks
+
+-- | The never-applicable attack.
+failingAttack :: Attack a
+failingAttack = empty
+
+-- | The attack that always applies and leaves the transaction unchanged.
+doNothingAttack :: Attack ()
+doNothingAttack = return ()
+
+-- | The "attack" that returns the current 'MockChainSt'
+mcstAttack :: Attack MockChainSt
+mcstAttack = Attack $ \mcst skel -> [(skel, mcst)]
 
--- * Constructing 'Attack's that return at most one modified transaction
+-- | The "attack" that obtains some value from the 'TxSkel'.
+viewAttack :: Is k A_Getter => Optic' k is TxSkel a -> Attack a
+viewAttack optic = Attack $ \_mcst skel -> [(skel, view optic skel)]
 
--- | The simplest way to make an attack from an optic: Try to apply a given
--- function of type @a -> Maybe a@ to all foci of an optic, modifying all foci
--- that return @Just@. This attack returns a singleton list of the modified
--- 'TxSkel' if at least one modification was successful, i.e. if at least one of
--- the foci evaluates to @Just@; otherwise it returns @[]@.
+-- | The attack that sets a certain value in the 'TxSkel'.
+setAttack :: Is k A_Setter => Optic' k is TxSkel a -> a -> Attack ()
+setAttack optic newValue = Attack $ \_mcst skel -> [(set optic newValue skel, ())]
+
+-- | The attack that modifies a certain value in the 'TxSkel'.
+overAttack :: Is k A_Setter => Optic' k is TxSkel a -> (a -> a) -> Attack ()
+overAttack optic change = Attack $ \_mcst skel -> [(over optic change skel, ())]
+
+-- * Composing attacks
+
+-- | Turn a potentially failing attack into an always successful attack, that
+-- just leaves the 'TxSkel' unmodified if the original attack would have failed.
+--
+-- In cases where the original attack would have failed, this returns
+-- @Nothing@. If the original attack would have been applicable, this is
+-- signalled by wrapping the original attack's return value in a @Just@.
+tryAttack :: Attack a -> Attack (Maybe a)
+tryAttack (Attack f) = Attack $
+  \mcst skel -> case f mcst skel of
+    [] -> [(skel, Nothing)]
+    l -> second Just <$> l
+
+-- | This attack returns the transaction as it has been modified by now. Use
+-- this as a kind of savepoint, if you want execute some sequence of attacks of
+-- which you're not sure that they will lead to the right result. You can
+-- restore the savepoint after that sequence with 'restoreHereAttack'.
+saveHereAttack :: Attack TxSkel
+saveHereAttack = Attack $ \_mcst skel -> [(skel, skel)]
+
+-- | See the comment at 'saveHereAttack'.
+restoreHereAttack :: TxSkel -> Attack ()
+restoreHereAttack skel = Attack $ \_ _ -> [(skel, ())]
```

Comment:
How about `getTxSkel`/`setTxSkel` as opposed to `save`/`restore`? The latter feels like implying some internal state.